### PR TITLE
ICU-20467 replace LocaleMatcher implementation

### PR DESCRIPTION
**Pull request by markusicu**
- move the XLocaleMatcher code into LocaleMatcher
- load likely-subtags & distance data from new langInfo.res bundle
- move data builders to tools
- make genrb handle multi-line binary values

https://unicode-org.atlassian.net/browse/ICU-20467

### CODE REVIEW

**Comment by macchiati** [comment]
Lgtm. Don't see where to approve...

**Comment by markusicu** [comment]
Yoshito PTAL

I amended my commit with the ant dependencies changes. Please see the diffs via the link from the Jira-GitHub PR Checker Bot.

Looks like I will need to amend further with a new icudata.jar.

**Comment by markusicu** [comment]
Yoshito: PTAL again

Had to rebase, resolve the icudata.jar conflict, and change one line in LikelySubtagsBuilder.java. No other changes.

Unfortunately, because of the rebase the cross-force-push diffs are large.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/locale/XLikelySubtags.java [hunk 1]
```diff
@@ -14,30 +22,93 @@
...
+            return new Data(languageAliases, regionAliases, trie, lsrs);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) { return true; }
+            if (!getClass().equals(other.getClass())) { return false; }
+            Data od = (Data)other;
+            return
+                    languageAliases.equals(od.languageAliases) &&
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/locale/XLikelySubtags.java]: Same as above - `Objects.equals(...)`

File: icu4j/main/classes/core/src/com/ibm/icu/util/LocaleMatcher.java [hunk 2]
```diff
@@ -54,880 +63,958 @@
...
+    public static final class Result {
+        private final ULocale desiredULocale;
+        private final ULocale supportedULocale;
+        private final Locale desiredLocale;
+        private final Locale supportedLocale;
+        private final int desiredIndex;
+        private final int supportedIndex;
+
+        private Result(ULocale udesired, ULocale usupported,
+                Locale desired, Locale supported,
```
Comment (yumaoka) [inline icu4j/main/classes/core/src/com/ibm/icu/util/LocaleMatcher.java]: Do we really need to store Locale in parallel to ULocale here?

File: icu4j/main/classes/core/src/com/ibm/icu/impl/locale/LocaleDistance.java [hunk 3]
```diff
@@ -54,22 +62,100 @@
...
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) { return true; }
+            if (!getClass().equals(other.getClass())) { return false; }
+            Data od = (Data)other;
+            return Arrays.equals(trie, od.trie) &&
+                    Arrays.equals(regionToPartitionsIndex, od.regionToPartitionsIndex) &&
+                    Arrays.equals(partitionArrays, od.partitionArrays) &&
+                    paradigmLSRs.equals(od.paradigmLSRs) &&
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/locale/LocaleDistance.java]: Neither is ever null, and this .equals() is only called from one test case.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/locale/XLikelySubtags.java (hunk 1 above)
Comment (markusicu) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/locale/XLikelySubtags.java]: ditto

File: icu4j/main/classes/core/src/com/ibm/icu/util/LocaleMatcher.java (hunk 2 above)
Comment (markusicu) [reply icu4j/main/classes/core/src/com/ibm/icu/util/LocaleMatcher.java]: The constructor always makes supported Locales from the supported ULocales, so both versions just get their references copied here. If someone has a build-match-discard use case, we can revisit lazy-building one or the other version.

One of the desired objects is real (reference copied from the input), the other is null. Their getters convert on the fly -- but I expect that if you match one type you won't normally ask to get the other. I considered making the getters cache the conversion result but then I would have to synchronize on the still-logically-immutable object.

**Pull request merged**